### PR TITLE
feat: Add userInfoGET endpoint

### PR DESCRIPTION
**Pull request by anku255**
## Summary of change

This PR adds the `userInfoGET` API  in the OAuth2 recipe.

## Related issues

-   https://github.com/supertokens/supertokens-node/issues/880

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [ ] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `lib/ts/version.ts`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [ ] Changes to the version if needed
    -   In `package.json`
    -   In `package-lock.json`
    -   In `lib/ts/version.ts`
-   [ ] Had run `npm run build-pretty`
-   [ ] Had installed and ran the pre-commit hook
-   [ ] If new thirdparty provider is added,
    -   [ ] update switch statement in `recipe/thirdparty/providers/configUtils.ts` file, `createProvider` function.
    -   [ ] add an icon on the user management dashboard.
-   [ ] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If have added a new web framework, update the `add-ts-no-check.js` file to include that
-   [ ] If added a new recipe / api interface, then make sure that the implementation of it uses NON arrow functions only (like `someFunc: function () {..}`).
-   [ ] If added a new recipe, then make sure to expose it inside the recipe folder present in the root of this repo. We also need to expose its types.
-   [ ] If added a new entry point, then make sure that it is importable by adding it to the `exports` in `package.json`

### CODE REVIEW

File: lib/ts/recipe/oauth2/api/userInfo.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+    if (user === undefined) {
+        sendNon200ResponseWithMessage(options.res, "Couldn't find any user associated with the access token", 401);
+        return true;
+    }
+
+    const response = await apiImplementation.userInfoGET({
+        accessTokenPayload,
+        user,
+        scopes: ((accessTokenPayload.scope as string) ?? "").split(" "),
```
Comment (porcellus) [inline lib/ts/recipe/oauth2/api/userInfo.ts:72]: Please check the type of `accessTokenPayload` above and return an error that says something like malformed access token.

File: lib/ts/recipe/oauth2/api/userInfo.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    const accessToken = authHeader.replace(/^Bearer /, "").trim();
+
+    let accessTokenPayload: JSONObject;
+
+    try {
+        accessTokenPayload = await validateOAuth2AccessToken(accessToken);
+    } catch (error) {
+        sendNon200ResponseWithMessage(options.res, "Invalid or expired OAuth2 access token!", 401);
```
Comment (porcellus) [inline lib/ts/recipe/oauth2/api/userInfo.ts:56]: Please add a TODO here (or check it yourself) to see if we need to set the `WWW-Authenticate` header in the response.

File: lib/ts/recipe/oauth2/constants.ts [hunk 3]
```diff
@@ -21,3 +21,4 @@ export const CONSENT_PATH = "/oauth2/consent";
 export const AUTH_PATH = "/oauth2/auth";
 export const TOKEN_PATH = "/oauth2/token";
 export const LOGIN_INFO_PATH = "/oauth2/login/info";
+export const USER_INFO_PATH = "/oauth2/userinfo";
```
Comment (porcellus) [inline lib/ts/recipe/oauth2/constants.ts:24]: I think this should be added to the openid discovery api response.

**Review by porcellus — CHANGES_REQUESTED** [review]
Please add userinfo builder functions for `role` and `permission` claims (registered by the `UserRoles` recipe)

File: lib/ts/recipe/oauth2/types.ts [hunk 4]
```diff
@@ -188,6 +188,15 @@ export type LoginInfo = {
     metadata?: Record<string, any> | null;
 };
 
+export type UserInfo = {
+    sub: string;
+    email?: string;
+    email_verified?: boolean;
+    phoneNumber?: string;
+    phoneNumber_verified?: boolean;
+    [key: string]: any;
```
[the remark this answers is not in the document]
Comment (anku255) [reply lib/ts/recipe/oauth2/types.ts:197]: done

File: lib/ts/recipe/oauth2/types.ts [hunk 5]
```diff
@@ -547,3 +564,9 @@ export type DeleteOAuth2ClientInput = {
 };
 
 export type PayloadBuilderFunction = (user: User, scopes: string[], userContext: UserContext) => Promise<JSONObject>;
+export type UserInfoBuilderFunction = (
+    user: User,
+    accessTokenPayload: JSONObject,
+    scopes: string[],
+    userContext: UserContext
+) => Promise<UserInfo>;
```
[the remark this answers is not in the document]
Comment (anku255) [reply lib/ts/recipe/oauth2/types.ts:572]: fixed.

File: lib/ts/recipe/oauth2/api/userInfo.ts (hunk 1 above)
Comment (anku255) [reply lib/ts/recipe/oauth2/api/userInfo.ts]: done

File: lib/ts/recipe/oauth2/api/userInfo.ts (hunk 2 above)
Comment (anku255) [reply lib/ts/recipe/oauth2/api/userInfo.ts]: done

File: lib/ts/recipe/oauth2/constants.ts (hunk 3 above)
Comment (anku255) [reply lib/ts/recipe/oauth2/constants.ts:24]: done

File: lib/ts/recipe/userroles/recipe.ts [hunk 6]
```diff
@@ -55,6 +56,51 @@ export default class Recipe extends RecipeModule {
...
             }
+
+            OAuth2Recipe.getInstanceOrThrowError().addUserInfoBuilderFromOtherRecipe(
+                async (user, _accessTokenPayload, scopes, tenantId, userContext) => {
+                    let userInfo: {
+                        roles?: string[];
+                        permissons?: string[];
+                    } = {};
+
+                    if (scopes.includes("roles")) {
```
Comment (porcellus) [inline lib/ts/recipe/userroles/recipe.ts:67]: the `permissions` scope should require the `roles` scope to be effective.

File: lib/ts/recipe/oauth2client/index.ts [hunk 7]
```diff
@@ -13,20 +13,25 @@
...
     static init = Recipe.init;
 
-    static async getAuthorisationRedirectURL(redirectURIOnProviderDashboard: string, userContext: UserContext) {
+    static async getAuthorisationRedirectURL(
+        redirectURIOnProviderDashboard: string,
+        userContext?: Record<string, any>
+    ) {
         const recipeInterfaceImpl = Recipe.getInstanceOrThrowError().recipeInterfaceImpl;
-        const providerConfig = await recipeInterfaceImpl.getProviderConfig({ userContext });
+        const providerConfig = await recipeInterfaceImpl.getProviderConfig({
```
Comment (porcellus) [inline lib/ts/recipe/oauth2client/index.ts:28]: these should be getting the same context object, so please extract `getUserContext(userContext)` to a constant and pass the same object to both recipe function calls.

**Review by porcellus — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/oauth2client/index.ts (hunk 7 above)
Comment (anku255) [reply lib/ts/recipe/oauth2client/index.ts:28]: Fixed.

File: lib/ts/recipe/userroles/recipe.ts (hunk 6 above)
Comment (anku255) [reply lib/ts/recipe/userroles/recipe.ts:67]: Isn't that already the case? 

The ` if (scopes.includes("permissions")) {` block is inside the `if (scopes.includes("roles")) {` block

Comment (porcellus) [reply lib/ts/recipe/userroles/recipe.ts:67]: oh I mistyped, it should *not* require it, the two can be used independently.

Comment (anku255) [reply lib/ts/recipe/userroles/recipe.ts]: Oh okay. I have fixed it.

**Review by porcellus — APPROVED** [review]

**Pull request merged**